### PR TITLE
Feature/issue 3 events

### PR DESCRIPTION
**Pull request by freshie**
This added the wait state and handles events
It also includes a lot of changes for more unit testable code and fixes many edge case when unit tests where added to the code.
The examples where moved into the example project and the common actions were moved up a level.

Closes #3

### CODE REVIEW

File: state-conductor-example/src/main/ml-data/state-conductor-flow/exception-flow.asl.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
       "Type": "Task",
       "End": true,
       "Comment": "runs a dhf 5 flow",
-      "Resource": "/state-conductor/actions/common/dhf/dhf5MakeEnvelopeStep.sjs",
+      "Resource": "/state-conductor/actions/custom/dhf/dhf5MakeEnvelopeStep.sjs",
```
Comment (aclavio) [inline state-conductor-example/src/main/ml-data/state-conductor-flow/exception-flow.asl.json]: Is this intentional to cause the exception this flow is testing?
The location of dh5MakeEnvelopStep.sjs is still in the /state-conductor/actions/common/dhf/ folder.

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs [hunk 2]
```diff
@@ -1,813 +1,1131 @@
...
+    xdmp.documentInsert(jobUri, job, {
+      collections: collections,
+      permissions: xdmp.defaultPermissions()
+    });
+  }, {
+    database: xdmp.database(STATE_CONDUCTOR_JOBS_DB)
+  });
+  
+  // add job metadata to the target document
+  addJobMetadata(uri, flowName, id); // prevents updates to the target from retriggering this flow
```
Comment (aclavio) [inline state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs]: I believe a change here may have been squashed by the merge.  Here's the code from the removed section:
```
// add job metadata to the target document (if one was passed)
  if (uri) {
    addJobMetadata(uri, flowName, id); // prevents updates to the target from retriggering this flow
  }
```

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs [hunk 3]
```diff
@@ -1,813 +1,1131 @@
...
+      });
+    }
+
+    // update the state status and provenence in the job doc
+    if (save) {
+      xdmp.nodeReplace(jobDoc.root, jobObj);
+    }
+  
+  } catch (err) {
+
```
Comment (aclavio) [inline state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs]: What could trigger this catch?  Seems unlikely since the condition module fallbacks are protected by their own try-catch.

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs [hunk 4]
```diff
@@ -1,813 +1,1131 @@
...
+  xdmp.trace(TRACE_EVENT, `resumeWaitingJob uri "${uri}"`);
+  xdmp.trace(TRACE_EVENT, `resumeWaitingJob flow "${flowName}"`);
+  xdmp.trace(TRACE_EVENT, `resumeWaitingJob flow state "${stateName}"`);
+
+  try {
+
+    // sanity check
+    if (FLOW_STATUS_WATING !== jobObj.flowStatus) {
+      return fn.error(null, 'INVALID-FLOW-STATUS', 'Cannot execute a flow that is not in the WAITING status');
```
Comment (aclavio) [inline state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs]: probably should replace "execute" with "resume" in the error description.

File: state-conductor-example/src/main/ml-data/state-conductor-flow/exception-flow.asl.json (hunk 1 above)
Comment (freshie) [reply state-conductor-example/src/main/ml-data/state-conductor-flow/exception-flow.asl.json]: This was a mistake I've fixed it

File: state-conductor-example/src/test/ml-modules/root/test/suites/StateConductorSuite/emmitEvent-test.sjs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+//random event jobs shouldnt be found
+assertion = sc.emmitEvent("theEndOfDays", 100, false);
+
+//they dont come back as as string so we have to convert them
+assertion = assertion.map(item => item.toString())
+
+assertions.push(test.assertEqual(assertion, [], "jobs shouldnt be found"))
+
+//assertions.push(test.assertEqual("working", assertion.flowStatus, "working flowStatus"))
```
[the remark this answers is not in the document]
Comment (freshie) [reply state-conductor-example/src/test/ml-modules/root/test/suites/StateConductorSuite/emmitEvent-test.sjs]: got catch. Missed when coping from qc. i've fixed it

File: state-conductor-example/src/test/ml-modules/root/test/suites/StateConductorSuite/executeStateByJobDoc-tests.sjs [hunk 6]
```diff
@@ -0,0 +1,214 @@
+'use strict';
+
+const sc = require('/state-conductor/state-conductor.sjs');
+const test = require('/test/test-helper.xqy');
+
+const assertions = [];
+let jobDoc, erorr, assertion;
```
[the remark this answers is not in the document]
Comment (freshie) [reply state-conductor-example/src/test/ml-modules/root/test/suites/StateConductorSuite/executeStateByJobDoc-tests.sjs]: fixed type O

File: state-conductor-example/src/test/ml-modules/root/test/suites/StateConductorSuite/startProcessingFlowByJobDoc-tests.sjs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    "flowState": "find-gender", 
+    "uri": "/data/test-doc3.json", 
+    "database": xdmp.database(), 
+    "modules": 1233456, 
+     "provenance": []
+  })
+ 
+  assertion = sc.startProcessingFlowByJobDoc(jobDoc, false);
+
+assertions.push(test.assertEqual("working", assertion.flowStatus, "unKnown database module"))
```
[the remark this answers is not in the document]
Comment (freshie) [reply state-conductor-example/src/test/ml-modules/root/test/suites/StateConductorSuite/startProcessingFlowByJobDoc-tests.sjs]: It doesn't produce an error because the startProcessingFlowByJobDoc doesnt use the modules. I was just adding a test to see what would happen if it was sent in like that and it should still work.

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs (hunk 4 above)
Comment (freshie) [reply state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs]: fixed error message

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/actions/common/emmit-event.sjs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+/**
+ * emmits an event
+ */
+
+const sc  = require('/state-conductor/state-conductor.sjs');
+
+function performAction(uri, options = {}) {
+
+  if (options.event) {
```
[the remark this answers is not in the document]
Comment (freshie) [reply state-conductor-modules/src/main/ml-modules/root/state-conductor/actions/common/emmit-event.sjs]: I agree we should have an error. I'll add it.

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs (hunk 2 above)
Comment (freshie) [reply state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs]: added this back in.

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs (hunk 3 above)
Comment (freshie) [reply state-conductor-modules/src/main/ml-modules/root/state-conductor/state-conductor.sjs]: This is to catch things like errors coming from inTerminalState function or there is no next step. IE the job was kicked off and then the flow was changed.

File: state-conductor-modules/src/main/ml-modules/root/state-conductor/actions/common/emmit-event.sjs (hunk 8 above)
[the remark this answers is not in the document]
Comment (freshie) [reply state-conductor-modules/src/main/ml-modules/root/state-conductor/actions/common/emmit-event.sjs]: I added the error
